### PR TITLE
Ensure event timer is active

### PR DESCRIPTION
**Pull request by cbartz**
### Overview

Ensure that the event timer is active on update_status. Add a log if the timer is not active.

### Rationale

We found on one deployment that the event timer was not active / could not be reactivated, resulting in only update status being run and no reconciliation.

### Juju Events Changes
The `update_status` hook now checks for the timer. There is a major refactoring going on in https://github.com/canonical/github-runner-operator/pull/199, the location of the check may be adjusted.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -154,11 +156,26 @@ def __init__(self, *args, **kargs) -> None:
             self.unit.status = ops.BlockedStatus(exc.msg)
             return
 
+        self._event_timer = EventTimer(self.unit.name)
+        if not self._event_timer.is_active(RECONCILE_RUNNERS_EVENT):
```
Comment (amandahla) [inline src/charm.py:160]: What happens if this command for checking if is active fails?

File: src/charm.py [hunk 2]
```diff
@@ -154,11 +156,26 @@ def __init__(self, *args, **kargs) -> None:
...
+            logger.warning(
+                "Reconciliation event timer is not activated - "
+                "this should only happen on the first charm execution after a deployment "
+                "on a new machine"
+            )
+            try:
+                self._set_reconcile_timer()
+            except TimerEnableError as ex:
+                logger.exception("Failed to start the event timer")
+                self.unit.status = BlockedStatus(
```
Comment (amandahla) [inline src/charm.py:170]: Out of curiosity: is there an alarm/notification for this case?

File: src/charm.py [hunk 3]
```diff
@@ -154,11 +156,31 @@ def __init__(self, *args, **kargs) -> None:
             self.unit.status = ops.BlockedStatus(exc.msg)
             return
 
+        self._event_timer = EventTimer(self.unit.name)
+        try:
+            reconcile_evt_timer_is_active = self._event_timer.is_active(RECONCILE_RUNNERS_EVENT)
```
Comment (yhaliaw) [inline src/charm.py:161]: Can `reconcile_evt_timer_is_active` be rename to `reconcile_event_time_is_active` or `reconcile_time_is_active`? Or something like that?
I feel `evt` a common shorthand for `event`.

File: src/charm.py [hunk 4]
```diff
@@ -154,11 +156,31 @@ def __init__(self, *args, **kargs) -> None:
...
+                    "Reconciliation event timer is not activated - "
+                    "this should only happen on the first charm execution after a deployment "
+                    "on a new machine."
+                )
+                try:
+                    self._set_reconcile_timer()
+                except TimerEnableError as ex:
+                    logger.exception("Failed to start the event timer")
+                    self.unit.status = BlockedStatus(
+                        (f"Failed to start timer for regular reconciliation checks: {ex}")
```
Comment (yhaliaw) [inline src/charm.py:176]: Is the `(`, `)` needed?

File: src/charm.py [hunk 5]
```diff
@@ -484,15 +514,11 @@ def _on_config_changed(self, _event: ConfigChangedEvent) -> None:
...
-                event_name="reconcile-runners",
-                interval=int(self.config["reconcile-interval"]),
-                timeout=int(self.config["reconcile-interval"]) - 1,
-            )
+            self._set_reconcile_timer()
         except TimerEnableError as ex:
             logger.exception("Failed to start the event timer")
             self.unit.status = BlockedStatus(
-                (f"Failed to start timer for regular reconciliation" f"checks: {ex}")
+                (f"Failed to start timer for regular reconciliation checks: {ex}")
```
Comment (yhaliaw) [inline src/charm.py:521]: Same as above

File: tests/unit/test_event_timer.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+from event_timer import EventTimer
+
+
+def test_is_active_true():
+    """
+    arrange: create an EventTimer object and mock exec command to return zero exit code
+    act: call is_active
+    assert: return True
+    """
+    evt = EventTimer(secrets.token_hex(16))
```
Comment (yhaliaw) [inline tests/unit/test_event_timer.py:15]: I feel `event` is clearer as a variable name

File: tests/unit/test_event_timer.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    assert evt.is_active(secrets.token_hex(16))
```
Comment (yhaliaw) [inline tests/unit/test_event_timer.py:16]: Maybe `evt.is_active(event_name=secrets.token_hex(16))`  makes it clearer that you are using a generated string as stub event name.
Same applies to other usage of `secrets.token_hex` in this file.

File: src/charm.py (hunk 1 above)
Comment (cbartz) [reply src/charm.py]: Thx, nice catch.

File: src/charm.py (hunk 2 above)
Comment (cbartz) [reply src/charm.py]: No, not at the moment. One could try to create an alert based on the log. But unfortunately there seems to be a bug in the grafana agent / loki charm that does not maintain the correct juju topology (so only log alerts can be defined for the grafana agent).

**Comment by cbartz** [comment]
Moved the check to the `update_status` hook. Reason: If placed in init, the code will be called before the install hook (first hook is `runner_storage_attached`), and the code would enable the timer that starts the service directly, and since the charm is not yet prepared (e.g. repo policy is not installed) it will fail and timeout after the reconciliation interval (this will resolve itself but still delay the installation for 10 minutes).

**Review by merkata — DISMISSED** [review]
lgtm

**Review by amandahla — DISMISSED** [review]
LGTM

**Review by amandahla — DISMISSED** [review]
LGTM

**Review by merkata — DISMISSED** [review]

**Review by amandahla — APPROVED** [review]
Nice!

**Review by merkata — APPROVED** [review]

**Pull request merged**